### ISSUE TITLE
SMTPSenderFactory has unintended side effect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10372]
|<img alt="mjs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mjs reported|
|-|-|
|Trac ID|trac#540|
|Type|defect|
|Created|2004-03-05 09:01:38Z|

<details><summary>Searchable metadata</summary>

```
trac-id__540 540
type__defect defect
reporter__mjs mjs
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1078477298000000 1078477298000000
changetime__1477011447504199 1477011447504199
version__ 
owner__ 
cc__exarkun cc__spiv cc__mjs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mjs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mjs commented|
|-|-|

```
#!html
<pre>
SMTPSenderFactory uses and modifies its toEmail argument directly without making
a copy first. This can have some unexpected side effects when using a constant
list of recipients as input.

eg.
from twisted.protocols import smtp
from twisted.internet import reactor

FROM_ADDR = 'from@domain.com'
TO_ADDRS = ['rcpt1@...', 'rcpt2@...']
TO_HOST = 'x.x.x.x'

def sendsomething():
    smtp.sendmail(TO_HOST, FROM_ADDR, TO_ADDRS, 'fooo')

sendsomething()
reactor.callLater(10, sendsomething)
reactor.run()

---

When the second call to sendsomething() occurs TO_ADDRS == [] and the email will
not be sent as intended (no recipients specified).

Workaround is to always copy TO_ADDRS before calling smtp.sendmail() but the
side-effect is not obvious and may not always be noticed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Resolved in r11825.  smtp.py now iterates over the to addresses, instead of
popping from the list.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc commented|
|-|-|

In changeset 01611cba886c3f9946bafacc9c028ff1e2592f53
```
#!CommitTicketReference repository="" revision="01611cba886c3f9946bafacc9c028ff1e2592f53"
Merge pull request #540 from twisted/8079-no-gmpy

Author: hawkowl
Reviewer: rodrigc
Fixes: #8079 

Remove gmpy support
```
